httpx/_utils.py

```diff
@@ -309,6 +310,37 @@ def unquote(value: str) -> str:
     return value[1:-1] if value[0] == value[-1] == '"' else value
 
 
+def guess_content_type(filename: typing.Optional[str]) -> typing.Optional[str]:
+    if filename:
+        return mimetypes.guess_type(filename)[0] or "application/octet-stream"
+    return None
+
+
+def get_filelike_length(stream: typing.IO) -> int:
+    """
+    Given a file-like stream object, efficiently compute and return its length
+    in number of bytes.
+    """
+    try:
+        # Is it an actual file?
+        fd = stream.fileno()
+    except OSError:
+        # No... is it something that supports random access, like `io.BytesIO`?
+        if not hasattr(stream, "seekable") or not stream.seekable():  # pragma: nocover
+            # Not even that? Sorry, we're doomed...
+            raise
```

Comment:
We might need to think about how we handle this case a bit more I guess. I don't recall off the top of my head if individual multipart file fields *have* to include a content-length, or if they support framing the data when the content length is unkown up front?

Reply:
Right, so yeah the content is delimited by the `boundary` markers, so files *without* a content-length header are perfectly okay. Eg in the MDN example here https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition

```
POST /test.html HTTP/1.1
Host: example.org
Content-Type: multipart/form-data;boundary="boundary"

--boundary
Content-Disposition: form-data; name="field1"

value1
--boundary
Content-Disposition: form-data; name="field2"; filename="example.txt"

value2
--boundary--
```

I *guess* we ought to include the Content-Length where possible, but it's feasible that our policy could alternately be "let's *not* include the Content-Length". Perhaps it'd be good to take a steer on that from some prior art. Eg. what does the streaming multipart implementation in the requests toolbelt do?... https://toolbelt.readthedocs.io/en/latest/uploading-data.html#streaming-multipart-data-encoder

Reply:
For the purposes of *this* function, I guess we'll want to return an `Optional[int]`, and return `None` when we cannot determine the content length.

Reply:
Ah interesting about `Content-Disposition`… I actually took heavy inspiration from this blog post: [Streaming Multipart Requests](http://khanlou.com/2018/11/streaming-multipart-requests/), and it also actually doesn't use the `Content-Disposition: form-data; ...` format too.

I'll try that out and do some validation against some live servers too. :-)

Reply:
See https://github.com/encode/httpx/pull/857#issuecomment-598063395: looks like in practice `Content-Length` is very much needed.

So my intuition here would be to fail loudly in case we can't compute the `Content-Length`.

What kind of exception would we like to throw? Surely bubbling up `OSError` isn't great UX. Would this be a new subclass of `StreamError`?

Reply:
Now stale, as we now handle this special 0.01%-of-cases situation by reading the entire file in memory as a fallback.